x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/services/custom/request/CustomRequest.java

```diff
@@ -0,0 +1,165 @@
+import org.elasticsearch.xpack.inference.external.request.HttpRequest;
+import org.elasticsearch.xpack.inference.external.request.Request;
+import org.elasticsearch.xpack.inference.services.custom.CustomModel;
+import org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings;
+import org.elasticsearch.xpack.inference.services.settings.SerializableSecureString;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.inference.common.JsonUtils.toJson;
+import static org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings.REQUEST_CONTENT;
+import static org.elasticsearch.xpack.inference.services.custom.CustomServiceSettings.URL;
+
+public class CustomRequest implements Request {
+    private static final String QUERY = "query";
+    private static final String INPUT = "input";
+
+    private final URI uri;
+    private final ValidatingSubstitutor jsonPlaceholderReplacer;
+    private final ValidatingSubstitutor stringPlaceholderReplacer;
+    private final CustomModel model;
+
+    public CustomRequest(String query, List<String> input, CustomModel model) {
+        this.model = Objects.requireNonNull(model);
+
+        var stringOnlyParams = new HashMap<String, String>();
+        addStringParams(stringOnlyParams, model.getSecretSettings().getSecretParameters());
+        addStringParams(stringOnlyParams, model.getTaskSettings().getParameters());
+
+        var jsonParams = new HashMap<String, String>();
+        addJsonStringParams(jsonParams, model.getSecretSettings().getSecretParameters());
+        addJsonStringParams(jsonParams, model.getTaskSettings().getParameters());
+
+        if (query != null) {
+            jsonParams.put(QUERY, toJson(query, QUERY));
+        }
+
+        addInputJsonParam(jsonParams, input, model.getTaskType());
+
+        jsonPlaceholderReplacer = new ValidatingSubstitutor(jsonParams, "${", "}");
+        stringPlaceholderReplacer = new ValidatingSubstitutor(stringOnlyParams, "${", "}");
+        uri = buildUri();
+    }
+
+    private static void addStringParams(Map<String, String> stringParams, Map<String, ?> paramsToAdd) {
+        for (var entry : paramsToAdd.entrySet()) {
+            if (entry.getValue() instanceof String str) {
+                stringParams.put(entry.getKey(), str);
+            } else if (entry.getValue() instanceof SerializableSecureString serializableSecureString) {
+                stringParams.put(entry.getKey(), serializableSecureString.getSecureString().toString());
+            } else if (entry.getValue() instanceof SecureString secureString) {
+                stringParams.put(entry.getKey(), secureString.toString());
+            }
+        }
+    }
+
+    private static void addJsonStringParams(Map<String, String> jsonStringParams, Map<String, ?> params) {
```

Comment:
Fields like the request body need to be a valid json object so we'll convert the values into json